// linkar_framework/Linkar.Functions/ASCII_Chars.js
    exports.SOH = 0x01;
    exports.SOH_chr = '\u0001';
    exports.SOH_str = "\u0001";

    exports.STX = 0x02;
    exports.STX_chr = '\u0002';
    exports.STX_str = "\u0002";

    exports.ETX = 0x03;
    exports.ETX_chr = '\u0003';
    exports.ETX_str = "\u0003";

    exports.EOT = 0x04;
    exports.EOT_chr = '\u0004';
    exports.EOT_str = "\u0004";

    exports.DC1 = 0x11;
    exports.DC1_chr = '\u0011';
    exports.DC1_str = "\u0011";

    exports.DC2 = 0x12;
    exports.DC2_chr = '\u0012';
    exports.DC2_str = "\u0012";

    exports.DC3 = 0x13;
    exports.DC3_chr = '\u0013';
    exports.DC3_str = "\u0013";

    exports.SUB = 0x1A;
    exports.SUB_chr = '\u001A';
    exports.SUB_str = "\u001A";

    exports.US = 0x1F;
    exports.US_chr = '\u001F';
    exports.US_str = "\u001F";

    exports.DC4 = 0x14;
    exports.DC4_chr = '\u0014';
    exports.DC4_str = "\u0014";

    exports.FS = 0x1C;
    exports.FS_chr = '\u001C';
    exports.FS_str = "\u001C";
    
    exports.RS = 0x1E;
    exports.RS_chr = '\u001E';
    exports.RS_str = "\u001E";

    exports.TAB = 0x09;
    exports.TAB_chr = '\t';
    exports.TAB_str = "\t";

    exports.LF = 0x0A;
    exports.LF_chr = '\n';
    exports.LF_str = "\n";
    
    exports.CR = 0x0D;
    exports.CR_chr = '\r';
    exports.CR_str = "\r";
